drivers/soft_spi/soft_spi.c

```diff
@@ -0,0 +1,202 @@
+ * @brief       Software SPI implementation
+ *
+ * @author      Markus Blechschmidt <Markus.Blechschmidt@haw-hamburg.de>
+ */
+
+#include "assert.h"
+#include "soft_spi.h"
+#include "periph/spi.h"
+#include "periph/gpio.h"
+#include "soft_spi_params.h"
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static bool spi_bus_is_valid(spi_t bus)
+{
+    unsigned int spi_num = (unsigned int) bus;
+
+    if (sizeof(spi_config) / sizeof(spi_config[0]) < spi_num || spi_num < 0) {
+        return false;
+    }
+    return true;
+}
+
+void spi_init(spi_t bus)
+{
+    /* This is left empty because the periph_init function will call this for
+       number of hardware spi ports */
+}
+
+void spi_init_pins(spi_t bus)
+{
+    soft_spi_t *curr_soft_spi;
+
+    assert(spi_bus_is_valid(bus));
+    curr_soft_spi = &soft_spi_config[(unsigned int) bus];
+    /* check that miso is not mosi is not clk*/
+    assert(curr_soft_spi->mosi_pin != curr_soft_spi->miso_pin);
+    assert(curr_soft_spi->mosi_pin != curr_soft_spi->clk_pin);
+    assert(curr_soft_spi->miso_pin != curr_soft_spi->clk_pin);
+    /* mandatory pins */
+    assert((GPIO_UNDEF != curr_soft_spi->mosi_pin) || (GPIO_UNDEF != curr_soft_spi->miso_pin));
+    assert(GPIO_UNDEF != curr_soft_spi->clk_pin);
+    /* check and setup mode related things */
+    assert(curr_soft_spi->spi_mode >= SPI_MODE_0);
+    assert(curr_soft_spi->spi_mode <= SPI_MODE_3);
+    /* initialize clock pin */
+    gpio_init(curr_soft_spi->clk_pin, GPIO_OUT);
+    /* initialize optional pins */
+    if (GPIO_UNDEF != curr_soft_spi->mosi_pin) {
+        gpio_init(curr_soft_spi->mosi_pin, GPIO_OUT);
+        gpio_clear(curr_soft_spi->mosi_pin);
+    }
+    if (GPIO_UNDEF != curr_soft_spi->miso_pin) {
+        gpio_init(curr_soft_spi->miso_pin, GPIO_IN);
+    }
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    if (!spi_bus_is_valid(bus)) {
+        return SPI_NODEV;
+    }
+    if (GPIO_UNDEF == cs) {
+        return SPI_NOCS;
+    }
+    /* unselect chip */
+    gpio_init(cs, GPIO_OUT);
+    gpio_set(cs);
+    return SPI_OK;
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    soft_spi_t *curr_soft_spi;
+
+    assert(spi_bus_is_valid(bus));
+    curr_soft_spi = &soft_spi_config[(unsigned int) bus];
+
+    if (!curr_soft_spi->initialized) {
+        mutex_init(&(curr_soft_spi->lock));
+    }
+    mutex_lock(&(curr_soft_spi->lock));
+
+    if (mode < SPI_MODE_0 || SPI_MODE_3 < mode) {
+        return SPI_NOMODE;
+    }
+    curr_soft_spi->spi_mode = mode;
+    switch (mode) {
+        case SPI_MODE_0:
+        case SPI_MODE_1:
+            /* CPOL=0 */
+            gpio_clear(curr_soft_spi->clk_pin);
+            break;
+        case SPI_MODE_2:
+        case SPI_MODE_3:
+            /* CPOL=1 */
+            gpio_set(curr_soft_spi->clk_pin);
+            break;
+    }
+    /* initialize pins */
+    if (!curr_soft_spi->initialized) {
+        spi_init_pins(bus);
+    }
+    spi_init_cs(bus, cs);
```

Comment:
The programmer might use an already initialized software spi bus with a new CS pin.